rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -222,6 +73,15 @@ class Publisher : public PublisherBase
   virtual ~Publisher()
   {}
 
+  mapped_ring_buffer::MappedRingBufferBase::SharedPtr
+  make_mapped_ring_buffer(size_t size) const override
+  {
+    return mapped_ring_buffer::MappedRingBuffer<
+      MessageT,
+      typename Publisher<MessageT, Alloc>::MessageAlloc
+    >::make_shared(size, this->get_allocator());
```

Comment:
Why was it necessary to move this here? Isn't this the reason that you now have a circular dependency that needed to be broken by moving lots of code around in the headers?